File: tests/vectors/verify_vectors.c
// Reference verifier for PQ-NAS QR-Auth v4 test vectors
// Requires: libsodium + OpenSSL (linked via pqnas_core)
//
// Run (from repo root):
//   ./build/bin/verify_vectors

#include <sodium.h>
#include <openssl/sha.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "qrauth_v4.h"

// -------------------- File helpers --------------------

static unsigned char *read_file(const char *path, size_t *out_len) {
    FILE *f = fopen(path, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    unsigned char *buf = (unsigned char*)malloc((size_t)sz + 1);
    if (!buf) { fclose(f); return NULL; }
    size_t n = fread(buf, 1, (size_t)sz, f);
    fclose(f);
    buf[n] = 0;
    if (out_len) *out_len = n;
    return buf;
}

// base64url no padding -> bytes
static int b64url_decode(const char *s, unsigned char *out, size_t out_max, size_t *out_len) {
    return sodium_base642bin(
        out, out_max,
        s, strlen(s),
        NULL,
        out_len,
        NULL,
        sodium_base64_VARIANT_URLSAFE_NO_PADDING
    );
}

static void b64url_encode(const unsigned char *in, size_t in_len, char *out, size_t out_max) {
    sodium_bin2base64(out, out_max, in, in_len, sodium_base64_VARIANT_URLSAFE_NO_PADDING);
}

// -------------------- Markdown extraction --------------------

static char *extract_fenced(const char *md, const char *header) {
    // Find "## <header>", then next fenced block
    char needle[256];
    snprintf(needle, sizeof(needle), "## %s", header);
    const char *p = strstr(md, needle);
    if (!p) return NULL;

    const char *a = strstr(p, "```");
    if (!a) return NULL;
    a += 3;

    const char *b = strstr(a, "```");
    if (!b) return NULL;

    size_t n = (size_t)(b - a);
    while (n > 0 && (a[0] == '\n' || a[0] == '\r' || a[0] == ' ' || a[0] == '\t')) { a++; n--; }
    while (n > 0 && (a[n-1] == '\n' || a[n-1] == '\r' || a[n-1] == ' ' || a[n-1] == '\t')) n--;

    char *out = (char*)malloc(n + 1);
    if (!out) return NULL;
    memcpy(out, a, n);
    out[n] = 0;
    return out;
}

static char *extract_server_pk(const char *md) {
    // Locate "## Server key", then fenced block after "public key"
    const char *p = strstr(md, "## Server key");
    if (!p) return NULL;

    p = strstr(p, "public key");
    if (!p) return NULL;

    const char *a = strstr(p, "```");
    if (!a) return NULL;
    a += 3;

    const char *b = strstr(a, "```");
    if (!b) return NULL;

    size_t n = (size_t)(b - a);
    while (n > 0 && (a[0] == '\n' || a[0] == '\r' || a[0] == ' ' || a[0] == '\t')) { a++; n--; }
    while (n > 0 && (a[n-1] == '\n' || a[n-1] == '\r' || a[n-1] == ' ' || a[n-1] == '\t')) n--;

    char *out = (char*)malloc(n + 1);
    if (!out) return NULL;
    memcpy(out, a, n);
    out[n] = 0;
    return out;
}

// -------------------- INVALID hash sanity check --------------------

static void sha256(const unsigned char *data, size_t len, unsigned char out[32]) {
    SHA256(data, len, out);
}

// -------------------- main --------------------

int main(void) {
    if (sodium_init() < 0) {
        fprintf(stderr, "libsodium init failed\n");
        return 2;
    }

    size_t md_len = 0;
    unsigned char *md = read_file("pqnas_qrauth_v4_test_vectors.md", &md_len);
    if (!md) {
        fprintf(stderr, "Missing pqnas_qrauth_v4_test_vectors.md in repo root\n");
        return 2;
    }

    char *server_pk_b64 = extract_server_pk((const char*)md);
    char *req_token = extract_fenced((const char*)md, "req_token");
    char *proof_token = extract_fenced((const char*)md, "proof_token");

    if (!server_pk_b64 || !req_token || !proof_token) {
        fprintf(stderr, "Failed to parse vectors file\n");
        free(md);
        return 2;
    }

    // IMPORTANT: markdown fences may wrap lines; tokens must be whitespace-free
    qr_strip_ws_inplace(server_pk_b64);
    qr_strip_ws_inplace(req_token);
    qr_strip_ws_inplace(proof_token);

    unsigned char server_pk[64];
    size_t server_pk_len = 0;
    if (b64url_decode(server_pk_b64, server_pk, sizeof(server_pk), &server_pk_len) != 0 || server_pk_len != 32) {
        fprintf(stderr, "Bad server public key\n");
        free(md);
        return 2;
    }

    // VALID
    if (qr_verify_req_token(req_token, server_pk) != QR_OK) {
        printf("VALID: FAIL (req_token server signature)\n");
        free(md);
        return 1;
    }
    if (qr_verify_proof_token(proof_token, req_token, server_pk) != QR_OK) {
        printf("VALID: FAIL (proof verification)\n");
        free(md);
        return 1;
    }
    printf("VALID: PASS\n");

    // INVALID (hash sanity)
    size_t inv_len = 0;
    unsigned char *inv = read_file("pqnas_qrauth_v4_test_vector_invalid.md", &inv_len);
    if (inv) {
        const char *invs = (const char*)inv;
        char *tampered = extract_fenced(invs, "Tampered req_token (used in signing)");
        char *wronghash = extract_fenced(invs, "Wrong req_hash_b64");

        if (!tampered || !wronghash) {
            printf("INVALID: SKIP (could not parse invalid vector file)\n");
        } else {
            qr_strip_ws_inplace(tampered);
            qr_strip_ws_inplace(wronghash);

            unsigned char wh[32];
            sha256((const unsigned char*)tampered, strlen(tampered), wh);
            char wh_b64[256];
            b64url_encode(wh, 32, wh_b64, sizeof(wh_b64));

            if (strcmp(wh_b64, wronghash) != 0) {
                printf("INVALID: FAIL\n");
                free(inv);
                free(md);
                return 1;
            }

            unsigned char ch[32];
            sha256((const unsigned char*)req_token, strlen(req_token), ch);
            char ch_b64[256];
            b64url_encode(ch, 32, ch_b64, sizeof(ch_b64));

            if (strcmp(ch_b64, wronghash) == 0) {
                printf("INVALID: FAIL\n");
                free(inv);
                free(md);
                return 1;
            }

            printf("INVALID: PASS (expected mismatch detected)\n");
        }
        free(inv);
    } else {
        printf("INVALID: SKIP (file missing)\n");
    }

    // INVALID_PROOF (signature mismatch)
    size_t invp_len = 0;
    unsigned char *invp = read_file("pqnas_qrauth_v4_test_vectors_invalid_proof.md", &invp_len);
    if (invp) {
        char *proof_token_invalid = extract_fenced((const char*)invp, "proof_token_invalid");
        if (!proof_token_invalid) {
            printf("INVALID_PROOF: SKIP (could not parse invalid proof file)\n");
        } else {
            qr_strip_ws_inplace(proof_token_invalid);

            qr_err_t rc = qr_verify_proof_token(proof_token_invalid, req_token, server_pk);
            if (rc == QR_OK) {
                printf("INVALID_PROOF: FAIL (unexpectedly verified)\n");
                free(invp);
                free(md);
                return 1;
            } else {
                printf("INVALID_PROOF: PASS (phone signature mismatch as expected)\n");
            }
        }
        free(invp);
    } else {
        printf("INVALID_PROOF: SKIP (file missing)\n");
    }

    free(md);
    return 0;
}

File: tests/vectors/verify_vectors.py
#!/usr/bin/env python3
import base64
import hashlib
import json
import re
import sys
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def b64url_encode_nopad(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")


def b64url_decode_nopad(s: str) -> bytes:
    pad = "=" * ((4 - (len(s) % 4)) % 4)
    return base64.urlsafe_b64decode((s + pad).encode("ascii"))


def extract_fenced(md: str, header: str) -> str:
    """
    Extract content from:
    ## <header>
    ```
    ...
    ```
    """
    pat = rf"## {re.escape(header)}\s+```([\s\S]*?)```"
    m = re.search(pat, md)
    if not m:
        raise ValueError(f"Could not find fenced block for header: {header}")
    return m.group(1).strip()


def extract_server_pk(md: str) -> str:
    """
    Extract server public key (base64url) from the vectors markdown.

    We locate the "Server key" section, then grab the first ``` fenced block after it.
    This is robust against formatting differences (em-dashes, line wrapping, etc.).
    """
    import re

    # Find "Server key" heading
    m = re.search(r"##\s*Server key.*?\n", md, flags=re.IGNORECASE)
    if not m:
        raise ValueError("Could not find 'Server key' section")

    start = m.end()
    tail = md[start:]

    # First fenced block after that section
    m2 = re.search(r"```([\s\S]*?)```", tail)
    if not m2:
        raise ValueError("Could not find server public key fenced block after 'Server key' section")

    return m2.group(1).strip()


def verify_req_token(
    req_token: str,
    server_pk_b64: str,
    now: int | None = None,
    check_time: bool = False,
) -> dict:
    try:
        payload_b64, sig_b64 = req_token.split(".", 1)
    except ValueError:
        raise ValueError("req_token must be payload_b64.sig_b64")

    payload_bytes = b64url_decode_nopad(payload_b64)
    sig = b64url_decode_nopad(sig_b64)
    payload = json.loads(payload_bytes.decode("utf-8"))

    # Server signs SHA256(payload_bytes)
    digest = hashlib.sha256(payload_bytes).digest()
    server_pk = Ed25519PublicKey.from_public_bytes(
        b64url_decode_nopad(server_pk_b64)
    )
    server_pk.verify(sig, digest)

    if payload.get("typ") != "req" or payload.get("v") != 4:
        raise ValueError("req_payload typ/v mismatch")

    if check_time:
        if now is None:
            now = int(__import__("time").time())

        if payload.get("exp") is not None and now > int(payload["exp"]):
            raise ValueError("req_token expired")

        if payload.get("iat") is not None and now + 300 < int(payload["iat"]):
            raise ValueError("req_token iat too far in future")

    return payload


def verify_proof_token(proof_token: str, server_pk_b64: str, now: int | None = None) -> dict:
    try:
        proof_payload_b64, phone_sig_b64 = proof_token.split(".", 1)
    except ValueError:
        raise ValueError("proof_token must be payload_b64.sig_b64")

    proof_payload_bytes = b64url_decode_nopad(proof_payload_b64)
    phone_sig = b64url_decode_nopad(phone_sig_b64)
    proof = json.loads(proof_payload_bytes.decode("utf-8"))

    if proof.get("typ") != "proof" or proof.get("v") != 4:
        raise ValueError("proof typ/v mismatch")

    req_token = proof["req"]
    fingerprint_b64 = proof["fingerprint"]
    pk_b64 = proof["pk"]
    ts = int(proof["ts"])

    # Verify req_token (server-signed)
    _ = verify_req_token(req_token, server_pk_b64, check_time=False)

    # Fingerprint binding: fingerprint == b64url(SHA3-512(pk_raw))
    pk_raw = b64url_decode_nopad(pk_b64)
    fp_calc = b64url_encode_nopad(hashlib.sha3_512(pk_raw).digest())
    if fp_calc != fingerprint_b64:
        raise ValueError("fingerprint binding mismatch")

    # req_hash_b64 = b64url(SHA256(UTF8(req_token)))
    req_hash_b64 = b64url_encode_nopad(hashlib.sha256(req_token.encode("utf-8")).digest())

    # Signing message (exact lines, no trailing newline)
    msg = "DNAQR-V4\n" + req_hash_b64 + "\n" + fingerprint_b64 + "\n" + str(ts)
    prehash = hashlib.sha3_512(msg.encode("utf-8")).digest()

    # Phone signature (Ed25519 TEST) over *prehash bytes*
    phone_pk = Ed25519PublicKey.from_public_bytes(pk_raw)
    phone_pk.verify(phone_sig, prehash)

    # Optional: timestamp freshness window
    if now is None:
        now = int(__import__("time").time())
    if abs(now - ts) > 600:
        # don’t hard-fail test vectors if your clock differs; warn
        pass

    return proof


def main() -> int:
    root = Path(".")
    valid_path = root / "pqnas_qrauth_v4_test_vectors.md"
    invalid_path = root / "pqnas_qrauth_v4_test_vector_invalid.md"

    # INVALID PROOF should fail (signature mismatch)
    invalid_proof_path = root / "pqnas_qrauth_v4_test_vectors_invalid_proof.md"
    if invalid_proof_path.exists():
        invp_md = invalid_proof_path.read_text(encoding="utf-8")
        proof_token_invalid = extract_fenced(invp_md, "proof_token_invalid")
        try:
            verify_proof_token(proof_token_invalid, server_pk_b64)
            print("INVALID_PROOF: FAIL (unexpectedly verified)")
            return 1
        except Exception as e:
            print(f"INVALID_PROOF: PASS ({e})")
    else:
        print("INVALID_PROOF: SKIP (file missing)")


if __name__ == "__main__":
    raise SystemExit(main())
